File: exp3/Round Robin.java
import java.util.*;

class RoundRobin {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt(); // number of processes
        int[] bt = new int[n]; // burst time
        int[] rem_bt = new int[n]; // remaining time
        int[] wt = new int[n]; // waiting time
        int[] tat = new int[n]; // turnaround time

        for (int i = 0; i < n; i++) {
            bt[i] = sc.nextInt();
            rem_bt[i] = bt[i];
        }

        int quantum = sc.nextInt();

        int time = 0;

        while (true) {
            boolean done = true;

            for (int i = 0; i < n; i++) {
                if (rem_bt[i] > 0) {
                    done = false;

                    if (rem_bt[i] > quantum) {
                        time += quantum;
                        rem_bt[i] -= quantum;
                    } else {
                        time += rem_bt[i];
                        wt[i] = time - bt[i];
                        rem_bt[i] = 0;
                    }
                }
            }

            if (done) break;
        }

        for (int i = 0; i < n; i++) {
            tat[i] = bt[i] + wt[i];
        }

        System.out.println("Process\tWT\tTAT");
        for (int i = 0; i < n; i++) {
            System.out.println(i + "\t" + wt[i] + "\t" + tat[i]);
        }

        sc.close();
    }
}

File: exp3/RoundRobin.cpp
#include <iostream>
using namespace std;

int main() {
    cout << "===== Round Robin Scheduling =====\n\n";

    int n = 4, tq = 2;
    int bt[] = {5, 3, 8, 6};
    int rem[4], wt[4] = {0}, tat[4];

    for(int i=0;i<n;i++) rem[i]=bt[i];

    int t = 0;

    while(true) {
        bool done = true;

        for(int i=0;i<n;i++) {
            if(rem[i] > 0) {
                done = false;

                if(rem[i] > tq) {
                    t += tq;
                    rem[i] -= tq;
                } else {
                    t += rem[i];
                    wt[i] = t - bt[i];
                    rem[i] = 0;
                }
            }
        }

        if(done) break;
    }

    for(int i=0;i<n;i++)
        tat[i] = bt[i] + wt[i];

    cout << "Process\tBT\tWT\tTAT\n";
    for(int i=0;i<n;i++)
        cout << "P" << i+1 << "\t" << bt[i] << "\t" << wt[i] << "\t" << tat[i] << endl;

    return 0;
}